avogadro/qtplugins/surfaces/surfaces.cpp

```diff
@@ -183,69 +186,149 @@ void Surfaces::calculateSurface()
-    case VanDerWaals:
-        break;
-    case SolventAccessible:
-        radiusOffset = 1.4;
-        break;
-    case SolventExcluded:
-        probeRadius = 1.4;
-        break;
-  }
-
-  // cache Van der Waals radii
-  std::vector<double> radii(m_molecule->atomCount());
-  double max_radius = probeRadius;
-  for (size_t i = 0; i < radii.size(); i++) {
-    radii[i] = Core::Elements::radiusVDW(m_molecule->atomicNumber(i));
-    if (radii[i] > max_radius)
-      max_radius = radii[i];
-  }
+  QFuture future = QtConcurrent::run([=]() {
+    double probeRadius = 0.0;
+    switch (m_dialog->surfaceType()) {
+      case VanDerWaals:
+          break;
+      case SolventAccessible:
+      case SolventExcluded:
+          probeRadius = 1.4;
+          break;
+    }
 
-  double padding = max_radius + radiusOffset;
-  m_cube->setLimits(*m_molecule, m_dialog->resolution(), padding);
-
-  auto neighborPerceiver = Core::NeighborPerceiver(
-    m_molecule->atomPositions3d(), 2.0 * max_radius
-  );
-  Vector3i cubeSize = m_cube->dimensions();
-
-  std::vector<Vector3i> *indices = new std::vector<Vector3i>;
-  indices->reserve(cubeSize[0] * cubeSize[1] * cubeSize[2]);
-  for (size_t zi = 0; zi < cubeSize[2]; zi++)
-    for (size_t yi = 0; yi < cubeSize[1]; yi++)
-      for (size_t xi = 0; xi < cubeSize[0]; xi++)
-        indices->emplace_back(xi, yi, zi);
-
-  const float res = m_dialog->resolution();
-  const Vector3 min = m_cube->min();
-  const float mdist = probeRadius;
-
-  thread_local Array<Index> *neighbors = nullptr;
-  QFuture future = QtConcurrent::map(*indices, [=](Vector3i &in) {
-    Vector3 pos(in(0), in(1), in(2));
-    pos += Vector3(0.5, 0.5, 0.5);
-    pos *= res;
-    pos += min;
-
-    double minDistance = mdist;
-    if (neighbors == nullptr)
-      neighbors = new Array<Index>;
-    neighborPerceiver.getNeighborsInclusiveInPlace(*neighbors, pos);
-    for (Index neighbor: *neighbors) {
-      Vector3 neighborPos = m_molecule->atomPosition3d(neighbor);
-      double distance = (neighborPos - pos).norm();
-      distance -= radii[neighbor] + radiusOffset;
-      minDistance = std::min(minDistance, distance);
+    // first, make a list of all atom positions and radii
+    Array<Vector3> atomPositions = m_molecule->atomPositions3d();
+    std::vector<std::pair<Vector3, double>> *atoms =
+      new std::vector<std::pair<Vector3, double>>(m_molecule->atomCount());
+    double max_radius = probeRadius;
+    for (size_t i = 0; i < atoms->size(); i++) {
+      (*atoms)[i].first = atomPositions[i];
+      (*atoms)[i].second = Core::Elements::radiusVDW(m_molecule->atomicNumber(i)) + probeRadius;
+      if ((*atoms)[i].second > max_radius)
+        max_radius = (*atoms)[i].second;
     }
 
-    m_cube->setValue(in(0), in(1), in(2), minDistance);
+    double padding = max_radius + probeRadius;
+    m_cube->setLimits(*m_molecule, m_dialog->resolution(), padding);
+    Vector3i size = m_cube->dimensions();
+    for (int z = 0; z < size(2); z++)
```

Comment:
A related method in `Cube` could be useful for filling regions (e.g., the low resolution algorithm). Instead of the triple loop here, such a method could use a double loop with `std::fill` on the underlying `m_data` vector.